File: sdk-ts/src/sub/events.ts
import type { Transport } from '../transport.js';

export interface RawEvent {
  eventId: string;
  contractId: string;
  topic: string;
  ledger: number;
  ledgerCloseTs: number;
  txHash: string;
  payload: unknown;
  insertedAt: number;
}

export interface EventQuery {
  topic?: string;
  contract?: string;
  fromLedger?: number;
  toLedger?: number;
  limit?: number;
}

export class EventsApi {
  constructor(private readonly transport: Transport) {}

  async list(query: EventQuery = {}): Promise<RawEvent[]> {
    const res = await this.transport.request<{ events: RawEvent[] }>({
      path: '/v1/events',
      query: {
        topic: query.topic,
        contract: query.contract,
        from_ledger: query.fromLedger,
        to_ledger: query.toLedger,
        limit: query.limit,
      },
    });
    return res.events;
  }
}

File: sdk-py/noether_sdk/sub/events.py
from ..models import RawEvent
from ..transport import Transport


class EventsApi:
    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    async def list(
        self,
        *,
        topic: str | None = None,
        contract: str | None = None,
        from_ledger: int | None = None,
        to_ledger: int | None = None,
        before_ts: int | None = None,
        limit: int | None = None,
    ) -> list[RawEvent]:
        """Raw decoded contract events captured by the indexer.

        before_ts is a cursor: only events with a ledger close time strictly
        below it are returned.
        """
        params = {
            "topic": topic,
            "contract": contract,
            "from_ledger": from_ledger,
            "to_ledger": to_ledger,
            "before_ts": before_ts,
            "limit": limit,
        }
        body = await self._transport.request("GET", "/v1/events", params=params)
        return [RawEvent.model_validate(e) for e in body.get("events", [])]
